utils/src/main/java/software/amazon/awssdk/utils/async/FlatteningSubscriber.java

```diff
@@ -68,34 +100,149 @@ public void cancel() {
 
     @Override
     public void onNext(Iterable<U> nextItems) {
-        synchronized (lock) {
-            currentBatch = StreamSupport.stream(nextItems.spliterator(), false)
-                                        .collect(Collectors.toCollection(LinkedList::new));
-            fulfillDemand();
+        nextItems.forEach(item -> {
+            Validate.notNull(item, "Items must not be null.");
+            allItems.add(item);
+        });
+        upstreamDemand.decrementAndGet();
+        handleStateUpdate();
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        onErrorFromUpstream = throwable;
+        handleStateUpdate();
+    }
+
+    @Override
+    public void onComplete() {
+        onCompleteCalledByUpstream = true;
+        handleStateUpdate();
+    }
+
+    /**
+     * This is invoked after each downstream request or upstream onNext, onError or onComplete.
+     */
+    private void handleStateUpdate() {
+        // Anything that happens after this if statement and before we set handlingStateUpdate to false is guaranteed to only
+        // happen on one thread. For that reason, we should only invoke onNext, onComplete or onError within that block.
+        if (!handlingStateUpdate.compareAndSet(false, true)) {
+            return;
+        }
+
+        try {
+            // If we've already called onComplete or onError, don't do anything.
+            if (terminalCallMadeDownstream) {
+                return;
+            }
+
+            // Call onNext, onComplete and onError as needed based on the current subscriber state.
+            handleOnNextState();
+            handleOnCompleteState();
+            handleOnErrorState();
+        } catch (Error e) {
+            throw e;
+        } catch (Throwable e) {
+            // This should not happen. Perhaps our downstream was naughty?
+            die(e);
+        } finally {
+            handlingStateUpdate.set(false);
+        }
+
+        // It's possible we had an important state change between when we decided to release the state update flag, and we
+        // actually released it. If that seems to have happened, handle that state change on this thread, because another
+        // thread is not guaranteed to come around and do so.
+        boolean itemQueueIsEmpty = allItems.isEmpty();
+
+        boolean onNextNeeded = !itemQueueIsEmpty && downstreamDemand.get() > 0;
+        boolean onCompleteNeeded = itemQueueIsEmpty && onCompleteCalledByUpstream;
+        boolean onErrorNeeded = onErrorFromUpstream != null;
+
```

Comment:
Are these checks strictly necessary here? Would it simplify the impl if it just called `handleStateUpdate()` unconditionally?

Reply:
That would loop forever, right?

Reply:
Oops forgot it didn't have a terminal condition other than the first CAS. Thanks I think the update is easier to read